### ISSUE TITLE
Conjunction of two disjunctions cause incorrect errors

### ISSUE THREAD

**Issue by ericanderson** [Microsoft/TypeScript#30648]
**TypeScript Version:**  3.3

**Search Terms:**
* Conjunction

**Code**

```ts
interface Small {
    small: true,
    callbackWhenSmallIsFalsy?: undefined;
}

interface NotSmall {
    small?: false;
    callbackWhenSmallIsFalsy: () => void;
}

interface Green {
    green?: false
    numberWhenGreenTrue?: undefined
}

interface NotGreen {
    green: true
    numberWhenGreenTrue: number;
}

type SmallProps = Small | NotSmall;
type GreenProps = Green | NotGreen;

type Props = SmallProps & GreenProps;

const try1: Props = {
    green: true,
    numberWhenGreenTrue: 5,
    small: true
} // OK

const try2: Props = {
    green: false,
    numberWhenGreenTrue: 5,
    small: true
} // Unexpected Error: 'callbackWhenSmallIsFalsy' is missing...
// Expected Error: 5 is not assignable to undefined for 'numberWhenGreenTrue'

 const try3: Props = {
    green: false,
    small: true
} // OK

 const try4: Props = {
    green: false,
    small: true,
    callbackWhenSmallIsFalsy: () => 5
} // Acceptable Error: Types of property 'small' are incompatible.
// TS can't know which of the conditions to use, so this is okay. NTH: It lists multiple possible errors
```

**Expected behavior:**
in `try3` I don't expect the error to be about `callbackWhenSmallIsFalsy`. I would expect an error about `green` or `numberWhenGreenTrue`

**Actual behavior:**
When the conjunction is evaluated as false, it appears that the compiler tries to force an error into the first part of the conjunction, when the falsehood is from the second part of the conjunction.

**Playground Link:** 
http://www.typescriptlang.org/play/#src=interface%20Small%20%7B%0A%20%20%20%20small%3A%20true%2C%0A%20%20%20%20callbackWhenSmallIsFalsy%3F%3A%20undefined%3B%0A%7D%0A%0Ainterface%20NotSmall%20%7B%0A%20%20%20%20small%3F%3A%20false%3B%0A%20%20%20%20callbackWhenSmallIsFalsy%3A%20()%20%3D%3E%20void%3B%0A%7D%0A%0Ainterface%20Green%20%7B%0A%20%20%20%20green%3F%3A%20false%0A%20%20%20%20numberWhenGreenTrue%3F%3A%20undefined%0A%7D%0A%0Ainterface%20NotGreen%20%7B%0A%20%20%20%20green%3A%20true%0A%20%20%20%20numberWhenGreenTrue%3A%20number%3B%0A%7D%0A%0Atype%20SmallProps%20%3D%20Small%20%7C%20NotSmall%3B%0Atype%20GreenProps%20%3D%20Green%20%7C%20NotGreen%3B%0A%0Atype%20Props%20%3D%20SmallProps%20%26%20GreenProps%3B%0A%0Aconst%20try1%3A%20Props%20%3D%20%7B%0A%20%20%20%20green%3A%20true%2C%0A%20%20%20%20numberWhenGreenTrue%3A%205%2C%0A%20%20%20%20small%3A%20true%0A%7D%20%2F%2F%20OK%0A%0Aconst%20try2%3A%20Props%20%3D%20%7B%0A%20%20%20%20green%3A%20false%2C%0A%20%20%20%20numberWhenGreenTrue%3A%205%2C%0A%20%20%20%20small%3A%20true%0A%7D%20%2F%2F%20Unexpected%20Error%3A%20'callbackWhenSmallIsFalsy'%20is%20missing...%0A%2F%2F%20Expected%20Error%3A%205%20is%20not%20assignable%20to%20undefined%20for%20'numberWhenGreenTrue'%0A%0A%20const%20try3%3A%20Props%20%3D%20%7B%0A%20%20%20%20green%3A%20false%2C%0A%20%20%20%20small%3A%20true%0A%7D%20%2F%2F%20OK%0A%0A%20const%20try4%3A%20Props%20%3D%20%7B%0A%20%20%20%20green%3A%20false%2C%0A%20%20%20%20small%3A%20true%2C%0A%20%20%20%20callbackWhenSmallIsFalsy%3A%20()%20%3D%3E%205%0A%7D%20%2F%2F%20Acceptable%20Error%3A%20Types%20of%20property%20'small'%20are%20incompatible.%0A%2F%2F%20TS%20can't%20know%20which%20of%20the%20conditions%20to%20use%2C%20so%20this%20is%20okay.%20NTH%3A%20It%20lists%20multiple%20possible%20errors

**Related Issues:**

**Comment by RyanCavanaugh** [comment]
I believe the root cause is that we only apply know how to apply one discriminant at a time when fishing for the best candidate target of a union during error reporting.

**Comment by jgilchrist** [comment]
I believe [this playground link](https://www.typescriptlang.org/play/#src=interface%20Discriminant1_False%20%7B%20discriminant_1%3A%20false%20%7D%0D%0Ainterface%20Discriminant1_True%20%7B%20discriminant_1%3A%20true%2C%20extra_field_1%3A%20null%20%7D%0D%0A%0D%0Atype%20Discriminant1%20%3D%20Discriminant1_True%20%7C%20Discriminant1_False%3B%0D%0A%0D%0A%0D%0Ainterface%20Discriminant2_False%20%7B%20discriminant_2%3A%20false%20%7D%0D%0Ainterface%20Discriminant2_True%20%7B%20discriminant_2%3A%20true%2C%20extra_field_2%3A%20null%20%7D%0D%0A%0D%0Atype%20Discriminant2%20%3D%20Discriminant2_True%20%7C%20Discriminant2_False%3B%0D%0A%0D%0A%0D%0Atype%20FullType%20%3D%20(Discriminant1_True%20%7C%20Discriminant1_False)%20%26%20Discriminant2%3B%0D%0A%0D%0Alet%20erroneous_type%3A%20FullType%20%3D%20%7B%0D%0A%20%20%20%20discriminant_1%3A%20false%2C%0D%0A%20%20%20%20discriminant_2%3A%20true%2C%0D%0A%7D%0D%0A) from my stack overflow question [here](https://stackoverflow.com/questions/55636449/unexpected-type-error-when-using-typescripts-type-intersections-with-discrimina) demonstrates the same issue.

In that example, there are two discriminants, and each discriminant decides whether an extra field should  be present in the object.

```js
// discriminant_2 is true, so extra_field_2 should be included in the object
{ discriminant_1: false, discriminant_2: true }

// However, the error message says extra_field_1 is missing from the object
Type '{ discriminant_1: false; discriminant_2: true; }' is not assignable to type 'FullType'.
  Type '{ discriminant_1: false; discriminant_2: true; }' is not assignable to type 'Discriminant1_True & Discriminant2_True'.
    Property 'extra_field_1' is missing in type '{ discriminant_1: false; discriminant_2: true; }' but required in type 'Discriminant1_True'.
```

**Comment by jgilchrist** [comment]
I've been doing some debugging - disclaimer: this is my first foray into the TypeScript codebase so I might have misunderstood what's going on here. However I'm fairly confident that the issue is in `findMostOverlappyType`, which doesn't appear to consider discriminants at all. I hardcoded this function to return the actual 'most overlappy type' (as if we excluded types based on the discriminants) and I get the expected error message in that case.

From my reading of `findMostOverlappyType`, all it considers is the overlap of index types of the source and target types, so finding the most overlappy type for `{ discriminant_1: false, discriminant_2: true}` in

```js
{ discriminant_1: false, discriminant_2: false }
{ discriminant_1: false, discriminant_2:  true }
{ discriminant_1:  true, discriminant_2: false }
{ discriminant_1:  true, discriminant_2:  true }
```

will depend entirely on iteration order. In reality though, we should be able to consider types with matching discriminant values as more overlappy than the others.

If that's correct, I think augmenting `findMostOverlappyType` to not consider types with mismatched discriminant values as overlapping (or at least to prioritise types in which more discriminant values match) should get us the error message we expect.

**Comment by Vittly** [comment]
I think I have one more case:

```ts
interface IX {
    prop?: 'x-val';
}

interface IY {
    prop?: 'y-val';
}

type XY = IX | IY;

interface IXY {
    theme?: IX['prop'] | IY['prop'];
}

interface IB {
    text: string;
}

function make(flag: boolean) {
   const x: IB & XY = { // <--- type error
       prop: flag ? 'x-val': 'y-val',
       text: 'base'
    };

   return x;
}
```

Looks like `IB & XY` is `(IB & IX) | (IB & IY)`. I expected same result as `IB & IXY` preduce but it worked differently.

Type error:

```
const x: (IB & IX) | (IB & IY)
Type '{ prop: "x-val" | "y-val"; text: string; }' is not assignable to type '(IB & IX) | (IB & IY)'.
  Type '{ prop: "x-val" | "y-val"; text: string; }' is not assignable to type 'IB & IY'.
    Type '{ prop: "x-val" | "y-val"; text: string; }' is not assignable to type 'IY'.
      Types of property 'prop' are incompatible.
        Type '"x-val" | "y-val"' is not assignable to type '"y-val" | undefined'.
          Type '"x-val"' is not assignable to type '"y-val" | undefined'.
```
